lightning/src/ln/channelmanager.rs

```diff
@@ -695,22 +695,43 @@ macro_rules! handle_error {
 	}
 }
 
+macro_rules! convert_chan_err {
+	($self: ident, $err: expr, $short_to_id: expr, $channel: expr, $channel_id: expr) => {
+		match $err {
+			ChannelError::Ignore(msg) => {
+				(false, MsgHandleErrInternal::from_chan_no_close(ChannelError::Ignore(msg), $channel_id.clone()))
+			},
+			ChannelError::Close(msg) => {
+				log_trace!($self.logger, "Closing channel {} due to close-required error: {}", log_bytes!($channel_id[..]), msg);
+				if let Some(short_id) = $channel.get_short_channel_id() {
+					$short_to_id.remove(&short_id);
+				}
+				let shutdown_res = $channel.force_shutdown(true);
+				(true, MsgHandleErrInternal::from_finish_shutdown(msg, *$channel_id, shutdown_res, $self.get_channel_update(&$channel).ok()))
+			},
+			ChannelError::CloseDelayBroadcast(msg) => {
+				log_error!($self.logger, "Channel {} need to be shutdown but closing transactions not broadcast due to {}", log_bytes!($channel_id[..]), msg);
+				if let Some(short_id) = $channel.get_short_channel_id() {
+					$short_to_id.remove(&short_id);
+				}
+				let shutdown_res = $channel.force_shutdown(false);
+				(true, MsgHandleErrInternal::from_finish_shutdown(msg, *$channel_id, shutdown_res, $self.get_channel_update(&$channel).ok()))
+			}
+		}
+	}
+}
+
 macro_rules! break_chan_entry {
 	($self: ident, $res: expr, $channel_state: expr, $entry: expr) => {
 		match $res {
 			Ok(res) => res,
-			Err(ChannelError::Ignore(msg)) => {
-				break Err(MsgHandleErrInternal::from_chan_no_close(ChannelError::Ignore(msg), $entry.key().clone()))
-			},
-			Err(ChannelError::Close(msg)) => {
-				log_trace!($self.logger, "Closing channel {} due to Close-required error: {}", log_bytes!($entry.key()[..]), msg);
-				let (channel_id, mut chan) = $entry.remove_entry();
-				if let Some(short_id) = chan.get_short_channel_id() {
-					$channel_state.short_to_id.remove(&short_id);
+			Err(e) => {
+				let (drop, res) = convert_chan_err!($self, e, $channel_state.short_to_id, $entry.get_mut(), $entry.key());
```

Comment:
Note, you're loosing the panic! when `CloseDelayBroadcast` shouldn't show up in path where `break_chan_entry` is called. Though not sure if it's worthy to enrich the new `convert_chan_err` with a new boolean flag to panic when it's called from here... Likely more error-prone.

Reply:
True, but we now handle it, so I'm not sure it matters?

Reply:
Yeah one day we'll make those errors macro better :p